File: projects/Vibe_XASR/windows_build/src/VibeXASR.Windows/Refine/CloudRequestLog.cs
using System;
using System.Collections.Generic;

namespace VibeXASR.Windows.Refine;

/// <summary>One "recent request" record (troubleshooting + one-tap issue). Port of CloudRequestLog.Entry.</summary>
internal sealed class CloudReqEntry
{
    public Guid Id { get; } = Guid.NewGuid();
    public DateTime At { get; init; }
    public string Provider { get; init; } = "";
    public string BaseUrl { get; init; } = "";
    public string Model { get; init; } = "";
    public string Status { get; init; } = "";   // "ok" | "timeout" | "error" | "skipped"
    public int Ms { get; init; }
    public string Input { get; init; } = "";     // text sent to the model (original ASR)
    public string Output { get; init; } = "";    // model reply (or error message)
    public string Prompt { get; init; } = "";    // the actual prompt sent (placeholders filled)
}

/// <summary>Ring buffer of recent cloud-refine requests, for the settings page. Thread-safe: the refiner
/// writes from a background task, the UI reads on the UI thread. Keeps the latest 20. Port of CloudRequestLog.</summary>
internal sealed class CloudRequestLog
{
    public static readonly CloudRequestLog Shared = new();

    private readonly object _lock = new();
    private readonly List<CloudReqEntry> _entries = new();
    private const int Cap = 20;
    private bool _enabled = true;
    private string _pendingOriginal = "";

    /// <summary>Whether to record (user can turn off). Thread-safe.</summary>
    public bool Enabled
    {
        get { lock (_lock) return _enabled; }
        set { lock (_lock) _enabled = value; }
    }

    /// <summary>The raw ASR output (no pinyin/replacement/defiller — engine raw) used as the next record's
    /// input. TrayApp sets it before each refine; CloudRefiner reads it when logging.</summary>
    public string PendingOriginal
    {
        get { lock (_lock) return _pendingOriginal; }
        set { lock (_lock) _pendingOriginal = value; }
    }

    private static string Clip(string s, int n) => s.Length > n ? s.Substring(0, n) + "…" : s;

    public void Record(string provider, string baseUrl, string model, string status, int ms,
                       string input, string output, string prompt)
    {
        var e = new CloudReqEntry
        {
            At = DateTime.Now, Provider = provider, BaseUrl = baseUrl, Model = model, Status = status, Ms = ms,
            Input = Clip(input, 4000), Output = Clip(output, 4000), Prompt = Clip(prompt, 6000),
        };
        lock (_lock)
        {
            if (!_enabled) return;
            _entries.Insert(0, e);
            if (_entries.Count > Cap) _entries.RemoveRange(Cap, _entries.Count - Cap);
        }
    }

    public List<CloudReqEntry> Snapshot() { lock (_lock) return new List<CloudReqEntry>(_entries); }
    public void Clear() { lock (_lock) _entries.Clear(); }
}

File: projects/Vibe_XASR/macos_build/native/app/Sources/VibeIME/CloudRequestLog.swift
import Foundation

/// 云端润色「最近请求」记录,便于用户排查 + 一键提 issue。
/// 线程安全:CloudRefiner 在后台线程写,设置页 UI 在主线程读。环形缓冲,保留最近 100 条。
final class CloudRequestLog: @unchecked Sendable {
    static let shared = CloudRequestLog()

    struct Entry: Sendable {
        let id = UUID()
        let at: Date
        let provider: String   // 服务商 key(如 openai / aliyun / custN)
        let baseURL: String
        let model: String
        let status: String     // "ok" | "timeout" | "error"
        let ms: Int
        let input: String      // 送入大模型的文本(规则处理后)
        let output: String     // 大模型返回(失败时=错误信息)
        let prompt: String     // 实际发送的提示词(占位符已替换)
    }

    private let lock = NSLock()
    private var entries: [Entry] = []
    private var _enabled = true
    private var _pendingOriginal = ""
    private let cap = 20

    /// 下一条记录用的「原始 ASR」(X-ASR 引擎原始输出,不含同音字/替换/顺滑等任何规则)。
    /// AppDelegate 在每次润色前设置;CloudRefiner 记录时取它当 input。
    var pendingOriginal: String {
        get { lock.lock(); defer { lock.unlock() }; return _pendingOriginal }
        set { lock.lock(); defer { lock.unlock() }; _pendingOriginal = newValue }
    }
    private static func clip(_ s: String, _ n: Int) -> String { s.count > n ? String(s.prefix(n)) + "…" : s }

    /// 是否记录(用户可在设置里关)。线程安全。
    var enabled: Bool {
        get { lock.lock(); defer { lock.unlock() }; return _enabled }
        set { lock.lock(); defer { lock.unlock() }; _enabled = newValue }
    }

    func record(provider: String, baseURL: String, model: String, status: String, ms: Int,
                input: String, output: String, prompt: String) {
        let e = Entry(at: Date(), provider: provider, baseURL: baseURL, model: model, status: status, ms: ms,
                      input: Self.clip(input, 4000), output: Self.clip(output, 4000), prompt: Self.clip(prompt, 6000))
        lock.lock(); defer { lock.unlock() }
        guard _enabled else { return }
        entries.insert(e, at: 0)
        if entries.count > cap { entries.removeLast(entries.count - cap) }
    }
    func snapshot() -> [Entry] { lock.lock(); defer { lock.unlock() }; return entries }
    func clear() { lock.lock(); defer { lock.unlock() }; entries.removeAll() }
}
